File: strava-booster/download/activity-downloader.go
package download

import (
	"encoding/json"
	"fmt"
	"log"
	"math"
	"net/http"
	"net/url"
	"strconv"
	"time"
)

func Activities(from time.Time, auth string) []ActivityEntity {
	var activities []ActivityEntity

	var hasData = true
	var currentPage = 1

	fmt.Printf("Downloading activities since %s...\n", from.Format("2006-01-02"))
	for hasData == true {
		page := downloadPage(from, currentPage, auth)
		activities = append(activities, page.Activities...)

		totalPages := int(math.Ceil(float64(page.Total) / float64(page.PerPage)))
		fmt.Printf("Pages downloaded %d of %d\n", currentPage, totalPages)

		hasData = currentPage < totalPages
		currentPage++
	}

	return activities
}

func downloadPage(from time.Time, page int, auth string) ActivitiesPage {
	err, req := buildRequest(from, page, auth)

	client := &http.Client{}
	resp, err := client.Do(req)
	if err != nil {
		log.Fatalln(err)
	}

	var data ActivitiesPage
	err = json.NewDecoder(resp.Body).Decode(&data)
	if err != nil {
		log.Fatalln(err)
	}

	return data
}

func buildRequest(from time.Time, page int, auth string) (error, *http.Request) {
	stravaUrl, err := buildUrl(from, page)

	req, err := http.NewRequest("GET", stravaUrl.String(), nil)
	req.Header.Add("X-Requested-With", "XMLHttpRequest")
	req.Header.Add("Cookie", "_strava4_session="+auth)

	return err, req
}

func buildUrl(from time.Time, page int) (*url.URL, error) {
	stravaUrl, err := url.Parse("https://www.strava.com/athlete/training_activities")
	if err != nil {
		log.Fatal(err)
	}

	values := stravaUrl.Query()
	values.Add("start_date", from.Format("01/02/2006"))
	values.Set("activity_type", "Run")
	values.Set("per_page", strconv.Itoa(20))
	values.Set("page", strconv.Itoa(page))

	stravaUrl.RawQuery = values.Encode()

	return stravaUrl, err
}

File: src/main/kotlin/io/ysb/strava/booster/raw/ActivityDownloader.kt
package io.ysb.strava.booster.raw

import tools.jackson.databind.ObjectMapper
import org.springframework.http.HttpEntity
import org.springframework.http.HttpHeaders
import org.springframework.http.HttpMethod
import org.springframework.stereotype.Service
import org.springframework.web.client.RestTemplate
import java.time.Instant
import java.time.LocalDate
import java.time.ZoneOffset
import java.time.format.DateTimeFormatter
import java.util.Locale

@Service
class ActivityDownloader(
    private val restTemplate: RestTemplate,
    private val mapper: ObjectMapper
) {
    fun downloadActivities(from: LocalDate, auth: String): List<ActivityEntity> {
        val activities = mutableListOf<ActivityEntity>()
        var hasData = true
        var currentPage = 1

        println("Downloading activities since ${from.format(DateTimeFormatter.ofPattern("yyyy-MM-dd"))}...")
        while (hasData) {
            val page = downloadPage(from, currentPage, auth)

            val totalPages = (page.total / page.perPage.toDouble()).toInt()
            println("Pages downloaded $currentPage of $totalPages")

            activities.addAll(page.activities)
            page.activities.forEach { logActivity(it) }

            hasData = currentPage < totalPages && activities.none { notBefore(it, from) }
            currentPage++
        }

        return activities.filter { !notBefore(it, from) }
    }

    private fun logActivity(it: ActivityEntity) {
        val zdt = Instant.ofEpochSecond(it.startDateLocalRaw).atZone(ZoneOffset.UTC)
        val date = zdt.format(DateTimeFormatter.ofPattern("yyyy-MM-dd"))
        val day = zdt.format(DateTimeFormatter.ofPattern("EEEE", Locale.ENGLISH))
        println(listOf(
            it.activityTypeDisplayName,
            it.name,
            day,
            date,
            zdt.format(DateTimeFormatter.ofPattern("HH:mm")),
            it.distance,
            it.elapsedTimeRaw,
            it.movingTimeRaw,
            it.elevationGain,
            it.activityUrl
        ).joinToString("\t"))
    }

    private fun notBefore(it: ActivityEntity, from: LocalDate) =
        it.startDateLocalRaw < from.atStartOfDay().toEpochSecond(ZoneOffset.UTC)

    private fun downloadPage(from: LocalDate, page: Int, auth: String): ActivitiesPage {
        val url = buildUrl(from, page)
        val headers = HttpHeaders().apply {
            set("X-Requested-With", "XMLHttpRequest")
            set("Cookie", "_strava4_session=$auth")
        }

        val entity = HttpEntity<String>(headers)
        val response = restTemplate.exchange(url, HttpMethod.GET, entity, String::class.java)

        return mapper.readValue(response.body, ActivitiesPage::class.java)
    }

    private fun buildUrl(from: LocalDate, page: Int): String {
        val stravaUrl = "https://www.strava.com/athlete/training_activities"

        return "$stravaUrl?start_date=${from.format(DateTimeFormatter.ofPattern("MM/dd/yyyy"))}&activity_type=Run&per_page=20&page=$page"
    }
}
